### PR TITLE
[REVIEW]Fix out of index errors encountered with  sampling on out of index samples 

### PR DESCRIPTION
**Pull request by VibhuJawa**
THIS PR does the following

- [x] Ensure we dont sample on out of range values
Issue: https://github.com/rapidsai/cugraph/issues/2828
- [x] Add tests for the sampling error
- [x] Ensure all the DGL examples here pass
https://github.com/rapidsai/dgl/blob/6ece904c69687adcd35a5ea41d1f5ca4ea01c0e2/examples/cugraph-pytorch/cugraph-local/rgcn-hetero/README.MD
- [x] Reformat out the non class specific utilities in prepration for DGL graph service class

### CODE REVIEW

File: python/cugraph/cugraph/gnn/dgl_extensions/feature_storage.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+import dask_cudf
+import cupy as cp
+from cugraph.experimental import MGPropertyGraph
+
+
+class CuFeatureStorage:
+    """
+    Storage for node/edge feature data.
+    """
+
```
Comment (VibhuJawa) [inline python/cugraph/cugraph/gnn/dgl_extensions/feature_storage.py:25]: Removed out of the graph_store file

File: python/cugraph/cugraph/gnn/dgl_extensions/utils/sampling.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+        if len(df) == 0:
+            dlpack_d[k] = (None, None, None)
+        else:
+            dlpack_d[k] = (
+                df[src_n].to_dlpack(),
+                df[dst_n].to_dlpack(),
+                df[eid_n].to_dlpack(),
+            )
+
+    return dlpack_d
```
Comment (VibhuJawa) [inline python/cugraph/cugraph/gnn/dgl_extensions/utils/sampling.py:164]: Moved from cugraph_store and renamed to `create_dlpack_d`. 
 https://github.com/rapidsai/cugraph/blob/fbd5f204989364af9d30ab2b4d8ad72e276ced0a/python/cugraph/cugraph/gnn/graph_store.py#L586-L598

File: python/cugraph/cugraph/gnn/dgl_extensions/utils/sampling.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    else:
+        return src_type in edge_types
+
+
+def _convert_can_etype_s_to_tup(canonical_etype_s):
+    src_type, etype, dst_type = canonical_etype_s.split(",")
+    src_type = src_type[2:-1]
+    dst_type = dst_type[2:-2]
+    etype = etype[2:-1]
+    return (src_type, etype, dst_type)
```
Comment (VibhuJawa) [inline python/cugraph/cugraph/gnn/dgl_extensions/utils/sampling.py:149]: Moved from graph_store , see below: 
https://github.com/rapidsai/cugraph/blob/fbd5f204989364af9d30ab2b4d8ad72e276ced0a/python/cugraph/cugraph/gnn/graph_store.py#L668-L673

File: python/cugraph/cugraph/gnn/dgl_extensions/utils/sampling.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+def sample_multiple_sgs(
+    sgs,
+    sample_f,
+    start_list_d,
+    start_list_dtype,
+    edge_dir,
+    fanout,
+    with_replacement,
+):
```
Comment (VibhuJawa) [inline python/cugraph/cugraph/gnn/dgl_extensions/utils/sampling.py:72]: Moved from `graph_store`: 

https://github.com/rapidsai/cugraph/blob/fbd5f204989364af9d30ab2b4d8ad72e276ced0a/python/cugraph/cugraph/gnn/graph_store.py#L621-L658

File: python/cugraph/cugraph/gnn/dgl_extensions/utils/sampling.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+
+def sample_single_sg(
+    sg,
+    sample_f,
+    start_list,
+    start_list_dtype,
+    start_list_range,
+    fanout,
+    with_replacement,
+):
```
Comment (VibhuJawa) [inline python/cugraph/cugraph/gnn/dgl_extensions/utils/sampling.py:111]: Moved from: 

https://github.com/rapidsai/cugraph/blob/fbd5f204989364af9d30ab2b4d8ad72e276ced0a/python/cugraph/cugraph/gnn/graph_store.py#L601-L618

**Review by VibhuJawa — COMMENTED** [review]
Added comments to help reviewers on where each file is coming from

File: python/cugraph/cugraph/gnn/dgl_extensions/utils/sampling.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    # of the seed dtype is not same as the node dtype
+    start_list = start_list.astype(start_list_dtype)
+
+    # Filter start list by ranges
+    # to enure the seed is with in index values
+    # see below:
+    # https://github.com/rapidsai/cugraph/blob/branch-22.12/cpp/src/prims/per_v_random_select_transform_outgoing_e.cuh
+    start_list = start_list[
+        (start_list >= start_list_range[0]) & (start_list <= start_list_range[1])
+    ]
```
Comment (VibhuJawa) [inline python/cugraph/cugraph/gnn/dgl_extensions/utils/sampling.py:124]: This is the logic change in this PR to fix :  https://github.com/rapidsai/cugraph/issues/2828 . Also see test added in this PR

**Review by rlratzel — COMMENTED** [review]
Just a couple of minor questions/suggestions.

File: python/cugraph/cugraph/gnn/dgl_extensions/utils/add_data.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+            for col in feat_name_obj.keys():
+                current_cols = feat_name_obj[col]
+                # Handle strings too
+                if isinstance(current_cols, str):
+                    current_cols = [current_cols]
+                covered_columns = covered_columns + current_cols
+
+            if set(covered_columns) != set(columns):
+                raise ValueError(
+                    f"All the columns {columns} not covered in {columns} "
```
[the remark this answers is not in the document]
Comment (VibhuJawa) [reply python/cugraph/cugraph/gnn/dgl_extensions/utils/add_data.py:45]: I could have sworn i fixed this earlier. Anyways, fixed it here

**Review by rlratzel — APPROVED** [review]

**Review by jnke2016 — APPROVED** [review]

**Comment by rlratzel** [comment]
@gpucibot merge

**Pull request merged**